File: analytical_engine/java/graphx-on-graphscope/src/main/java/com/alibaba/graphscope/graph/impl/GraphXVertexIdManagerImpl.java
package com.alibaba.graphscope.graph.impl;

import com.alibaba.graphscope.conf.GraphXConf;
import com.alibaba.graphscope.ds.Vertex;
import com.alibaba.graphscope.fragment.IFragment;
import com.alibaba.graphscope.graph.GraphXVertexIdManager;
import com.alibaba.graphscope.utils.FFITypeFactoryhelper;
import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class GraphXVertexIdManagerImpl implements GraphXVertexIdManager {

    private static Logger logger = LoggerFactory.getLogger(GraphXVertexIdManagerImpl.class.getName());

    private GraphXConf conf;
    private Long[] oids;
    private long innerVerticesNum, fragVerticesNum;
    private Long2LongOpenHashMap oid2Lid;

    public GraphXVertexIdManagerImpl(GraphXConf conf) {
        this.conf = conf;
    }

    @Override
    public void init(IFragment<Long, Long, ?, ?> fragment) {
        //all frag vertices;
        Vertex<Long> vertex = FFITypeFactoryhelper.newVertexLong();
        for (long lid = 0; lid < fragment.getVerticesNum(); ++lid) {
            vertex.SetValue(lid);
            long oid = fragment.getId(vertex);
            oids[(int) lid] = oid;
            oid2Lid.put(oid, lid);
        }
        logger.info("Finish VertexId manager construction: fragId [{}], vertices [{}]", fragment.fid(), fragment.getVerticesNum());
        innerVerticesNum = fragment.getInnerVerticesNum();
        fragVerticesNum = fragment.getVerticesNum();
    }

    @Override
    public Long lid2Oid(Long lid) {
        return oids[lid.intValue()];
    }

    @Override
    public Long oid2Lid(Long oid){
        return oid2Lid.getOrDefault(oid, -1L);
    }

    @Override
    public long innerVerticesNum() {
        return innerVerticesNum;
    }

    @Override
    public long verticesNum() {
        return fragVerticesNum;
    }
}

File: analytical_engine/java/graphx-on-graphscope/src/main/scala/org/apache/spark/graphx/impl/graph/GraphXVertexIdManagerImpl.scala
package org.apache.spark.graphx.impl.graph

import com.alibaba.graphscope.ds.Vertex
import com.alibaba.graphscope.fragment.IFragment
import com.alibaba.graphscope.utils.FFITypeFactoryhelper
import it.unimi.dsi.fastutil.longs.Long2LongOpenHashMap
import org.apache.spark.graphx.GraphXConf
import org.apache.spark.graphx.traits.GraphXVertexIdManager
import org.apache.spark.internal.Logging


class GraphXVertexIdManagerImpl[VD,ED](val conf: GraphXConf[VD,ED], fragment: IFragment[Long,Long,_,_]) extends GraphXVertexIdManager with Logging{

  val innerVerticesNum: Long = fragment.getInnerVerticesNum
  val fragVerticesNum: Long = fragment.getVerticesNum
  val lid2Oids = new Array[Long](fragVerticesNum.toInt)
  private val oid2Lids: Long2LongOpenHashMap = new Long2LongOpenHashMap()
  val vertex: Vertex[Long] = FFITypeFactoryhelper.newVertexLong.asInstanceOf[Vertex[Long]]
  var lid = 0
  while (lid < fragVerticesNum) {
    vertex.SetValue(lid)
    val oid: Long = fragment.getId(vertex)
    lid2Oids(lid) = oid
    oid2Lids.put(oid, lid)
    lid += 1
  }
  log.info(s"Finish VertexId manager construction: fragId ${fragment.fid()}, vertices ${fragment.getVerticesNum}")


  override def lid2Oid(lid: Long): Long = {
    lid2Oids(lid.toInt)
  }

  override def oid2Lid(oid: Long): Long = {
    oid2Lids.getOrDefault(oid, -(1L))
  }

  override def getInnerVerticesNum: Long = {
    innerVerticesNum
  }

  override def getVerticesNum: Long = {
    fragVerticesNum
  }
}
